src/translator/quality_estimator.cpp

```diff
@@ -0,0 +1,209 @@
+    if (isspace(firstLetter)) {
+      ++subword.begin;
+      wordsFeatures.push_back({1, subwordScore, subwordScore});
+      wordByteRanges.push_back(subword);
+    } else {
+      // update last word byte range and word features
+
+      ByteRange& currentWord = wordByteRanges.back();
+      WordFeatures& currentWordFeatures = wordsFeatures.back();
+
+      currentWord.end = subword.end;
+      ++currentWordFeatures.numSubWords;
+      // incremental mean
+      currentWordFeatures.meanScore += (subwordScore - currentWordFeatures.meanScore) / currentWordFeatures.numSubWords;
+
+      if (currentWordFeatures.minScore > subwordScore) {
+        currentWordFeatures.minScore = subwordScore;
+      }
+    }
+  }
+
+  const float overallMean = sequence / subwordIdx;
+
+  return {wordByteRanges, wordsFeatures, overallMean};
+}
+
+std::vector<float> QualityEstimator::LogisticRegressor::vectorResult(const IntgemmMatrix& features) const {
+  float quant_mult = 1024.0f;
+
+  AlignedVector<int16_t> A_prepared(features.data.size());
+  AlignedVector<int16_t> B_prepared(coefficients_.data.size());
+
+  intgemm::Int16::PrepareA(features.data.begin(), A_prepared.begin(), quant_mult, features.rows, features.cols);
+  intgemm::Int16::PrepareB(coefficients_.data.begin(), B_prepared.begin(), quant_mult, coefficients_.rows,
+                           coefficients_.cols);
+
+  AlignedVector<float> modelScores(features.rows * coefficients_.cols);
+
+  intgemm::Int16::Multiply(
+      A_prepared.begin(), B_prepared.begin(), features.rows, features.cols, coefficients_.cols,
+      intgemm::callbacks::UnquantizeAndWrite(1.0f / (quant_mult * quant_mult), modelScores.begin()));
+
+  std::vector<float> wordQualityScores(features.rows);
+
+  for (int i = 0; i < features.rows; ++i) {
+    wordQualityScores[i] = modelScores[i * coefficients_.cols] + intercept_;
+  }
+
+  return wordQualityScores;
+}
+
+QualityEstimator::IntgemmMatrix QualityEstimator::LogisticRegressor::transformFeatures(
+    const std::vector<WordFeatures>& wordsFeatures, const float overallMean) const {
+  const intgemm::Index numWords = wordsFeatures.size();
+
+  QualityEstimator::IntgemmMatrix features(numWords, NumberOfFeatures, intgemm::Int16::tile_info.a_rows,
+                                           intgemm::Int16::tile_info.a_cols);
+
+  const auto getStds = [](const auto stds) { return stds != 0.0 ? stds : 1.0f; };
+
+  for (int i = 0; i < numWords; ++i) {
+    const WordFeatures& wordFeatures = wordsFeatures[i];
+    features.data[features.cols * i + 0] = (wordFeatures.meanScore - scale_.means[0]) / getStds(scale_.stds[0]);
+    features.data[features.cols * i + 1] = (wordFeatures.minScore - scale_.means[1]) / getStds(scale_.stds[1]);
+    features.data[features.cols * i + 2] = (wordFeatures.numSubWords - scale_.means[2]) / getStds(scale_.stds[2]);
+    features.data[features.cols * i + 3] = (overallMean - scale_.means[3]) / getStds(scale_.stds[3]);
+  }
```

Comment:
We have created the struct Matrix with the at(row, col) method since we believed it made the code easier to understand. We don't know if this would affect SIMD optimization. Do you think it's ok? Commit 96e36aa